yellow-paper/docs/circuits/private-kernel-initial.md

```diff
@@ -0,0 +1,258 @@
+
+> Note that the final transaction data is not deterministic for a given transaction request. The production of new notes, the destruction of notes, and various other values are likely to change based on the time and conditions when a transaction is being composed. However, the intricacies of implementation should not be a concern for the entity initiating the transaction.
+
+### Processing Private Function Call
+
+#### Ensuring the contract instance being called is deployed.
+
+It proves that the nullifier representing the contract exists in the contract tree.
+
+This nullifier is the contract address siloed with the address of a precompiled deployment contract.
+
+#### Ensuring the function being called exists in the contract.
+
+The contract address contains the contract class ID, which is a hash of the root of its function tree and additional values. This circuit leverages these characteristics to establish the validity of the function's association with the contract address.
+
+Each leaf of the function tree is a hash representing a function. The preimage includes:
+
+- Function data.
+- Hash of the verification key.
+- Hash of the function bytecode.
+
+To ensure the function's existence, the circuit executes the following steps:
+
+1. Computes the hash of the verification key.
+2. Calculates the function leaf: `hash(...function_data, vk_hash, bytecode_hash)`
+3. Derives the function tree root with the leaf and the specified sibling path.
+4. Computes the contract class ID using the function tree root and additional information.
+5. Generates the contract address using the contract class ID and other relevant details.
+6. Validates that the contract address matches the address specified in the private call data.
+
+#### Verifying the private function proof.
+
+It verifies that the private function was executed successfully with the provided proof data, verification key, and the public inputs of the private function circuit.
+
+#### Verifying the public inputs of the private function circuit.
+
+It ensures the private function circuit's intention by checking the following:
+
+- The contract address for each non-empty item in the following arrays must equal the current contract address:
+  - Note hash contexts.
+  - Nullifier contexts.
+  - L2-to-L1 message contexts.
+  - Read requests.
+- The portal contract address for each non-empty L2-to-L1 message context must equal the current portal contract address.
+- If the new contract contexts array is not empty, the contract address must equal the precompiled deployment contract address.
+- The historical data must match the one in the constant data.
+
+> Ensuring the alignment of the contract addresses is crucial, as it is later used to silo the value and to establish associations with values within the same contract.
+
+#### Verifying the call requests.
+
+For both private and public call requests initiated in the current function call, it ensures that for each request at index _i_:
```

Comment:
It would be useful to define the request index _i_ 

e.g. the request index is a unique identifier assigned to public and private call requests. The first call request Index value is 0. The index value increases by 1 for each subsequent call